go-controller/hybrid-overlay/pkg/windows/service/service.go

```diff
@@ -0,0 +1,111 @@
+// +build windows
+
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Based on https://github.com/kubernetes/kubernetes/blob/master/pkg/windows/service/service.go
+The reason for making a copy is that packages in kubernetes/kubernetes
+are not setup to be imported as modules without vendoring and pinning the sub
+components using the "require" directive. If more packages need to be imported
+then we can revisit importing this module too.
+*/
+
+package service
+
+import (
+	"os"
+	"time"
+
+	"k8s.io/klog"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+)
+
+var (
+	service *handler
+)
+
+type handler struct {
+	tosvc   chan bool
+	fromsvc chan error
+}
+
+// InitService is the entry point for running the daemon as a Windows
+// service. It returns an indication of whether it is running as a service;
+// and an error.
+func InitService(serviceName string) error {
+	h := &handler{
+		tosvc:   make(chan bool),
+		fromsvc: make(chan error),
+	}
+
+	service = h
+	var err error
+	go func() {
+		err = svc.Run(serviceName, h)
+		h.fromsvc <- err
+	}()
+
+	// Wait for the first signal from the service handler.
+	err = <-h.fromsvc
+	if err != nil {
+		return err
+	}
+	klog.Infof("Running %s as a Windows service!", serviceName)
+	return nil
+}
+
+func (h *handler) Execute(_ []string, r <-chan svc.ChangeRequest, s chan<- svc.Status) (bool, uint32) {
+	s <- svc.Status{State: svc.StartPending, Accepts: 0}
+	// Unblock initService()
+	h.fromsvc <- nil
+
```

Comment:
I am unsure of how the Execute calling works and how we can call c.RunContext(ctx, os.Args) in Execute() since I am not calling it anywhere in the code, I wouldn't be able to pass struct c along, are you suggesting we expose the struct? It would be helpful if you could expand on this idea.

Reply:
Execute is a golang construct that comes with the services package. There is a bit of documentation about it here: https://godoc.org/golang.org/x/sys/windows/svc#Handler

Basically, Execute is a callback (a.k.a "the handler function"), which name and parameters are hardcoded. The callback is magically called by golang whenever Run is called, and IIRC Run exits whenever the code exits the Execute callback. Additionally, IIRC, the value that Execute returns is forwarded to the Run and will be Run return value. Let me know if you have any additional questions on this!

Reply:
In that case we cannot pass the struct c in Execute() and we would need to expose it outside the package to be used in service/service.go is that right?

Reply:
Updating this thread with the content of the discussion we had offline...
You typically would want to get a pointer to "c" inside the "handler" structure, then set this "c" pointer somewhere before calling Run(); and then within the Execute() function, you would just access runContext() through h.c.runContext()

Reply:
> Instead of leaving the main() function do the "runContext", I would spin up a go routine here that would execute this code:
> 

In addition to my previous comment, won't this break the Linux hybrid-overlay? Or is this only Windows specific?

> ```
> if err := c.RunContext(ctx, os.Args); err != nil {
> 	klog.Exit(err)			klog.Exit(err)
> }		}
> ```
> 
> Then, I would move the initForOS() call that is currently in runHybridOverlay() back to main().

My previous comment applies here too.

> I would then skip the following code (currently in main() in hybrid-overlay-node.go) in case of a Windows service:
> 
> ```
> exitCh := make(chan os.Signal, 1)
> signal.Notify(exitCh,		signal.Notify(exitCh,
> 	syscall.SIGHUP,			syscall.SIGHUP,
> 	syscall.SIGINT,			syscall.SIGINT,
> 	syscall.SIGTERM,			syscall.SIGTERM,
> 	syscall.SIGQUIT)			syscall.SIGQUIT)
> defer func() {		defer func() {
> 	signal.Stop(exitCh)			signal.Stop(exitCh)
> 	cancel()			cancel()
> }()		}()
> go func() {		go func() {
> 	select {			select {
> 	case s := <-exitCh:			case s := <-exitCh:
> 		klog.Infof("Received signal %s. Shutting down", s)				klog.Infof("Received signal %s. Shutting down", s)
> 		cancel()				cancel()
> 	case <-ctx.Done():			case <-ctx.Done():
> 	}			}
> }()		}()
> ```
> 
> That is exactly what the Windows service manager is responsible for.

I agree this code needs skipping in the case of Windows though I am not sure how to figure out that we want to run as a Windows service at that point. One option is to say that hybrid-overlay-node.exe always needs to run as a service on Windows.